File: src/services/state.service.ts
import { AppointmentStatus } from '../types/appointment';
import { logger } from '../utils/logger';

export class StateService {
  // Set de Message IDs procesados recientemente para idempotencia estricta (evita reenvíos de webhook)
  private static processedMessageIds = new Set<string>();

  // Bloqueos en vuelo por teléfono para prevenir race conditions si el cliente envía mensajes ráfaga
  private static locksInFlight = new Set<string>();

  // Registro de últimos estados y respuestas para evitar bucles
  private static userLastInteractions = new Map<
    string,
    { status: AppointmentStatus; lastInteractionAt: number; count: number }
  >();

  /**
   * Verifica si un mensaje ya ha sido procesado (Idempotencia)
   */
  public static isMessageAlreadyProcessed(messageId: string): boolean {
    if (!messageId) return false;
    if (this.processedMessageIds.has(messageId)) {
      logger.warn({ messageId }, 'Mensaje duplicado detectado. Ignorando para mantener idempotencia.');
      return true;
    }
    this.processedMessageIds.add(messageId);

    // Mantener tamaño de caché controlado (máximo 10,000 IDs en memoria)
    if (this.processedMessageIds.size > 10000) {
      const first = this.processedMessageIds.values().next().value;
      if (first) this.processedMessageIds.delete(first);
    }
    return false;
  }

  /**
   * Adquiere un bloqueo atómico para un número de teléfono.
   * Evita condiciones de carrera si entran 2 webhooks simultáneos del mismo usuario.
   */
  public static async acquireLock(phone: string): Promise<boolean> {
    if (this.locksInFlight.has(phone)) {
      logger.warn({ phone }, 'Bloqueo activo: otra petición está procesando este teléfono actualmente');
      return false;
    }
    this.locksInFlight.add(phone);
    return true;
  }

  /**
   * Libera el bloqueo del teléfono
   */
  public static releaseLock(phone: string): void {
    this.locksInFlight.delete(phone);
  }

  /**
   * Registra interacción y detecta si el usuario está en un bucle repetitivo
   */
  public static registerInteraction(phone: string, newStatus: AppointmentStatus): { isSpamLoop: boolean } {
    const now = Date.now();
    const existing = this.userLastInteractions.get(phone);

    if (existing) {
      // Si interactúa más de 5 veces en menos de 2 minutos sobre el mismo estado terminal
      const diffMs = now - existing.lastInteractionAt;
      if (diffMs < 120000 && existing.status === newStatus && (newStatus === 'CONFIRMED' || newStatus === 'CANCELLED')) {
        existing.count += 1;
        existing.lastInteractionAt = now;
        if (existing.count >= 3) {
          logger.warn({ phone, status: newStatus, count: existing.count }, 'Detectado posible bucle de mensajes del usuario');
          return { isSpamLoop: true };
        }
      } else {
        existing.count = 1;
        existing.status = newStatus;
        existing.lastInteractionAt = now;
      }
    } else {
      this.userLastInteractions.set(phone, { status: newStatus, lastInteractionAt: now, count: 1 });
    }

    return { isSpamLoop: false };
  }

  /**
   * Determina si un estado es terminal y no debe aceptar mutaciones repetitivas
   */
  public static isTerminalStatus(status: AppointmentStatus): boolean {
    return status === 'CONFIRMED' || status === 'CANCELLED';
  }
}

File: python_app/state_service.py
import time
import logging
from types_models import AppointmentStatus

logger = logging.getLogger(__name__)

class StateService:
    _processed_message_ids = set()
    _locks_in_flight = set()
    _user_interactions = {}  # phone -> {"status": AppointmentStatus, "timestamp": float, "count": int}

    @classmethod
    def is_message_already_processed(cls, message_id: str) -> bool:
        if not message_id:
            return False
        if message_id in cls._processed_message_ids:
            logger.warning(f"Mensaje duplicado detectado: {message_id}. Ignorando para mantener idempotencia.")
            return True
        cls._processed_message_ids.add(message_id)

        # Evitar crecimiento ilimitado
        if len(cls._processed_message_ids) > 10000:
            cls._processed_message_ids.pop()
        return False

    @classmethod
    def acquire_lock(cls, phone: str) -> bool:
        if phone in cls._locks_in_flight:
            logger.warning(f"Bloqueo activo para {phone}: colisión prevenida.")
            return False
        cls._locks_in_flight.add(phone)
        return True

    @classmethod
    def release_lock(cls, phone: str) -> None:
        cls._locks_in_flight.discard(phone)

    @classmethod
    def register_interaction(cls, phone: str, status: AppointmentStatus) -> bool:
        """
        Retorna True si el usuario parece estar en un bucle repetitivo de spam.
        """
        now = time.time()
        existing = cls._user_interactions.get(phone)
        if existing:
            diff = now - existing["timestamp"]
            if diff < 120 and existing["status"] == status and status in [AppointmentStatus.CONFIRMED, AppointmentStatus.CANCELLED]:
                existing["count"] += 1
                existing["timestamp"] = now
                if existing["count"] >= 3:
                    logger.warning(f"Bucle de spam detectado para {phone} en estado {status}")
                    return True
            else:
                existing["count"] = 1
                existing["status"] = status
                existing["timestamp"] = now
        else:
            cls._user_interactions[phone] = {"status": status, "timestamp": now, "count": 1}
        return False

    @staticmethod
    def is_terminal_status(status: AppointmentStatus) -> bool:
        return status in [AppointmentStatus.CONFIRMED, AppointmentStatus.CANCELLED]
